## Fantasy Baseball Auction Value ##

__This notebook provides the content for today's lecture, as well as an introduction to Assignment 4. The details for Assignment 4 are provided at the end of the notebook.__

#### Due: Wednesday, March 20 by 3pm. ####

### Authors: Blaise Page and Ryan Turner ### 
#### Sabermetrics ####
#### Assignment 4 ####

Hello everyone, not Rhonda here for once (it's your CAs, Ashna and Brandon). As you've probably heard by now, we're going to be doing a fantasy baseball league in this class. The object of this assignment is to come up with some way for evaluating players. How you might ask? However you want we might reply. You've heard about WAR and wOBA and replacement value and all sorts of ways to evaluate players, and now we're going to turn it into a competition with (fake) money on it (and a prize at the end). 

#### How does fantasy baseball work?

Fantasy baseball assigns point values to different events for real like players, such as runs, homeruns, innings pitched, etc. You can see how your leagues scoring and rosters are set up in the section below under "You mentioned scoring?". At the end of the competition (near the end of the semester), whatever team has the most points in each league is the winner! Ashna and I will be playing as well to give you all some competition.

__The winning teams get to choose the type of cake we order on the last day of class.__

#### How do we pick our team?

In fantasy baseball, each player can only belong to one team (just like in real life). You will be selecting your team with a partner through an **auction draft**. In an auction draft, you'll start with \\$250. You'll take turns nominating a player, and then everyone will live bid on that player, just like a real life auction. The least a player can go for is \\$1, while the most is \\$250-(# roster spots - 1). You'll have to use your money wisely to build the best team possible, however you can still use waivers and trades to keep your roster competitive.

#### What are waivers?

Waivers are where any non-rostered player is at. You can pick up players from waivers, but you'll have to drop a player from your roster to do so.

#### You mentioned scoring?

| Batting Stat | Points | 
|--------------|--------|
| AB           | -1.0   |
| H            | 5.6    |
| 2B           | 2.9    |
| 3B           | 5.7    |
| HR           | 9.4    |
| BB           | 3.0    |
| HBP          | 3.0    |
| SB           | 1.9    |
| CS           | -2.8   |  
  
| Pitching Stat | Points |
|---------------|--------|
| IP            | 5      |
| K             | 2.0    |
| BB            | -3.0   |
| HBP           | -3.0   |
| HR            | -13    |
| SV            | 5.0    |
| HLD           | 4.0    |  
  
| Position | Roster Spots |
|----------|--------------|
| C        | 1            |
| 1B       | 1            |
| 2B       | 1            |
| 3B       | 1            |
| SS       | 1            |
| OF       | 3            |
| UTIL     | 1            |
| SP       | 5            |
| RP       | 2            |
| Bench    | 3            |  
  
A few notes about these settings. ESPN counts the "IP" stat as per out, so in the scoring settings it is 5/3=1.66, however per every other statkeeping reference, it's 5 points per IP. A 2B/3B/HR is added on to the value of a hit, so a double is worth 5.6+2.9=8.5 points. These weights are based on wOBA/FIP, adjusted to keep pitchers and batter on a reatively level playing field. OF is any outfield position. UTIL is any batter, doesn't matter the position (kind of like a DH).

# Assignment 4 - Fantasy Baseball #

### Join a league ###

The first step in this assignment is to join a fantasy league. There are two leagues you can join that are setup specifically for this class. Both have identcal settings, so pick one.   
http://fantasy.espn.com/baseball/league/join?leagueId=90287999&inviteId=471eac27-e559-4737-8f3e-6c13f054a6a1
http://fantasy.espn.com/baseball/league/join?leagueId=5441683&inviteId=cf19e2aa-38d9-4886-8aa6-9465aeb58369  
You'll need an ESPN account (free). In order to add your group-mate so that you can both control your team, go to "My Team" at the top of the page, and then under your team name, there's a "Add 2nd Owner" link. If you're having trouble with this, let one of Ashna or I know and we'll add you ourselves. 

### Design an approach to valuing players ###

Glad you asked. Now you get to put your real baseball skills to work. Your assignment is to come up with a way to value players. You can use FanGraphs projections and base them off of that. You can look at historical data and create your own prediction model. You can use some fancy neural network or machine learning algorithm if you want to go the extra mile. It's a pretty open ended question, just at the end you need to come up with some sort of a list that says "Player A is work \\$X" for every fantasy relevant player in the league. 

#### Not entirely sure where to get started? ####

That's fair, it's a pretty open ended task, especially if you've never done fantasy sports or an auction draft before. Here are some links to help you learn the ropes:  
  
[FanGraphs Fantasy Auction Calculator.](https://www.fangraphs.com/auctiontool.aspx) Now you can't just export this and turn it in, but it can be a helpful tool for knowing a general range for values.  
  
[A basic autcion draft overview.](https://757fantasyleague.wordpress.com/auction-draft-the-basics/) This can help you get more familiar with how an auction draft works and some of the basic strategies. You can't get every player that you want, so you need to budget you money.  
  
Personally, I would recommend coming up with a way to evaluate players based on the categories that we use for scoring. This should net you some expected points. Again, this can come from whatever math or form you like. Then worry about turning those point values into auction dollar values, and there are a number of ways to do that (you can do some Googling). 

#### How about some sweet hints?

No worries, in order to help even out the playing field a bit, here are some hints that can help you in the draft.  
1. Try to get other people to spend their money on players that you don't want but you think that other people will. You can use ESPN draft rankings you help find discrepencies between your rankings and the ones that other people will see to help with that.  
2. There's more than one strategy to building a team. You can blow all your money on some top tier talent and fill in the rest of your team with cheap players, or build a more balanced roster. Both can work.  
3. There's position scarcity in baseball (some positions are deeper than others), meaning that the worst 2B on a roster might put up 1.9 points a game while the worst 1B on a roster might put up 2.5. Establishing a replacement level *wink wink* for each position can help you guage which players are more valuable just becuase of the position they play. 
4. Fantasy value does not always equal real life value. We don't care about defense and players have to play to score points. There might be some great guy sitting on an MLB teams bench, but he's not going to give you fantasy points if he's only playing every third day. Here, counting stats matter.  
5. The league is only going to last a little over a month (the last day of points scoring will be 11:59pm on April 28th), meaning that prospects or players that won't be called up until midway through the season (Vlad Jr., Eloy) aren't realy worth drafting. You're welcome to ask if you don't know who those players are.

### Submit your code and writeup ###

The product that you produce for this assignment is a notebook that describes your approach, as well as the code that implements the approach. Your notebook needs to include the following sections and information:

1. Introduction - A summary of your general approach
2. Data and statistics - Describe the data and statistics that you used, and how they were used.
3. Results of your approach - This section needs to include the top 5 players that your approach identifies. You also need to select one player from your top 5 and provide an analysis of his rating and monetary value.
4. Limitations - What difficulties do you see with implementing your strategy in the draft?

The draft will occur in class on March 20. Your assignment needs to be submitted before class on that day. 

To make grading easier, please name the notebook *Assignment4_LastName1_LastName2*, where *LastName1* and *LastName2* are the last names of the team members. Yes, you will be working in teams of two for this assignment.

## Pitcher Selection Code ##

In [19]:
# pitcher selection

import pandas as pd
import numpy as np
from pybaseball import pitching_stats

pitch_stats = pitching_stats(2018)
pitch_stats.head(10)


Season              Name          Team   Age     W    L   ERA  WAR     G  \
41   2018.0      Jacob deGrom          Mets  30.0  10.0  9.0  1.70  8.5  32.0   
92   2018.0      Max Scherzer     Nationals  33.0  18.0  7.0  2.53  7.1  33.0   
91   2018.0  Justin Verlander        Astros  35.0  16.0  9.0  2.52  6.9  34.0   
57   2018.0        Chris Sale       Red Sox  29.0  12.0  4.0  2.11  6.8  27.0   
172  2018.0    Patrick Corbin  Diamondbacks  28.0  11.0  7.0  3.15  6.6  33.0   
133  2018.0       Gerrit Cole        Astros  27.0  15.0  5.0  2.88  6.5  32.0   
64   2018.0      Trevor Bauer       Indians  27.0  12.0  6.0  2.21  6.4  28.0   
74   2018.0        Aaron Nola      Phillies  25.0  17.0  6.0  2.37  5.9  33.0   
217  2018.0     Luis Severino       Yankees  24.0  19.0  8.0  3.39  5.8  32.0   
136  2018.0      Corey Kluber       Indians  32.0  20.0  7.0  2.89  5.7  33.0   

       GS    ...      wSL/C (pi)  wXX/C (pi)  O-Swing% (pi)  Z-Swing% (pi)  \
41   32.0    ...            2.25         NaN          0.367          0.661   
92   33.0    ...            1.85         NaN          0.355          0.666   
91   34.0    ...            0.48         NaN          0.336          0.656   
57   27.0    ...            2.20         NaN          0.363          0.602   
172  33.0    ...            2.12         NaN          0.379          0.621   
133  32.0    ...            0.53         NaN          0.315          0.650   
64   27.0    ...            2.78         NaN          0.314          0.609   
74   33.0    ...             NaN         NaN          0.354          0.599   
217  32.0    ...            1.33         NaN          0.323          0.649   
136  33.0    ...            3.20         NaN          0.353          0.620   

     Swing% (pi)  O-Contact% (pi)  Z-Contact% (pi)  Contact% (pi)  Zone% (pi)  \
41         0.518            0.524            0.804          0.708       0.513   
92         0.519            0.513            0.774          0.690       0.528   
91         0.509            0.557            0.785          0.716       0.541   
57         0.481            0.507            0.769          0.669       0.493   
172        0.471            0.469            0.863          0.667       0.381   
133        0.493            0.502            0.805          0.714       0.532   
64         0.451            0.455            0.853          0.704       0.463   
74         0.470            0.583            0.834          0.735       0.476   
217        0.496            0.557            0.835          0.750       0.530   
136        0.482            0.538            0.875          0.748       0.484   

     Pace (pi)  
41        21.9  
92        24.2  
91        27.0  
57        22.7  
172       24.0  
133       23.7  
64        23.6  
74        26.6  
217       22.0  
136       23.6  

[10 rows x 299 columns]

In [20]:
#print(pitch_stats['Relieving'])
list(pitch_stats)

['Season',
 'Name',
 'Team',
 'Age',
 'W',
 'L',
 'ERA',
 'WAR',
 'G',
 'GS',
 'CG',
 'ShO',
 'SV',
 'BS',
 'IP',
 'TBF',
 'H',
 'R',
 'ER',
 'HR',
 'BB',
 'IBB',
 'HBP',
 'WP',
 'BK',
 'SO',
 'GB',
 'FB',
 'LD',
 'IFFB',
 'Balls',
 'Strikes',
 'Pitches',
 'RS',
 'IFH',
 'BU',
 'BUH',
 'K/9',
 'BB/9',
 'K/BB',
 'H/9',
 'HR/9',
 'AVG',
 'WHIP',
 'BABIP',
 'LOB%',
 'FIP',
 'GB/FB',
 'LD%',
 'GB%',
 'IFFB%',
 'HR/FB',
 'IFH%',
 'BUH%',
 'Starting',
 'Start-IP',
 'Relieving',
 'Relief-IP',
 'RAR',
 'Dollars',
 'tERA',
 'xFIP',
 'WPA',
 '-WPA',
 '+WPA',
 'RE24',
 'REW',
 'pLI',
 'inLI',
 'gmLI',
 'exLI',
 'Pulls',
 'WPA/LI',
 'Clutch',
 'FB%',
 'FBv',
 'SL%',
 'SLv',
 'CT%',
 'CTv',
 'CB%',
 'CBv',
 'CH%',
 'CHv',
 'SF%',
 'SFv',
 'KN%',
 'KNv',
 'XX%',
 'PO%',
 'wFB',
 'wSL',
 'wCT',
 'wCB',
 'wCH',
 'wSF',
 'wKN',
 'wFB/C',
 'wSL/C',
 'wCT/C',
 'wCB/C',
 'wCH/C',
 'wSF/C',
 'wKN/C',
 'O-Swing%',
 'Z-Swing%',
 'Swing%',
 'O-Contact%',
 'Z-Contact%',
 'Contact%',
 'Zone%',
 'F-Strike%',
 'S

In [21]:
#print(len(pitch_stats)) # 784
from collections import OrderedDict

#starting_pitcher_dict = OrderedDict()
#relieving_pitcher_dict = OrderedDict()
pitcher_dict = OrderedDict()

# populate pitcher_dict with pitcher names as keys
for index, row in pitch_stats.iterrows():
    # if pitcher is a starter
    #if(row['Start-IP']>=row['Relief-IP']):
    #    starting_pitcher_dict[row['Name']] = {}
    # if pitcher is a reliever
    #else:
    #    relieving_pitcher_dict[row['Name']] = {}
    
    # make the master key the pitchers name
    pitcher_dict[row['Name']] = OrderedDict()
    # IP data
    pitcher_dict[row['Name']]['IP'] = row['IP']
    # K data
    pitcher_dict[row['Name']]['K'] = row['SO']
    # BB data
    pitcher_dict[row['Name']]['BB'] = row['BB']
    # HBP data
    pitcher_dict[row['Name']]['HBP'] = row['HBP']
    # HR data
    pitcher_dict[row['Name']]['HR'] = row['HR']
    # SV data
    pitcher_dict[row['Name']]['SV'] = row['SV']
    # HLD data
    pitcher_dict[row['Name']]['HLD'] = row['HLD']
    # WAR
    pitcher_dict[row['Name']]['WAR'] = row['WAR']
    
#print(pitcher_dict)
#for pitcher in pitcher_dict:
#    print(pitcher_dict[pitcher])

In [22]:
# Calculate each pitchers fantasy score as described in the writeup above
# IP=5; K=2; BB=-3; HBP=-3; HR=-13; SV=5; HLD=4
for pitcher in pitcher_dict:
    pitcher_dict[pitcher]['Fantasy_Score'] = 5*pitcher_dict[pitcher]['IP'] + 2*pitcher_dict[pitcher]['K'] + -3*pitcher_dict[pitcher]['BB'] + -3*pitcher_dict[pitcher]['HBP'] + -13*pitcher_dict[pitcher]['HR'] + 5*pitcher_dict[pitcher]['SV'] + 4*pitcher_dict[pitcher]['HLD']
    
#print(pitcher_dict)

In [23]:
# Calculate final score based on applying weights to both the Fantasy_Score and WAR
Fantasy_weight = .05
WAR_weight = 3.185

sum_scores = 0

for pitcher in pitcher_dict:
    pitcher_dict[pitcher]['Final_Score'] = Fantasy_weight*pitcher_dict[pitcher]['Fantasy_Score'] + WAR_weight*pitcher_dict[pitcher]['WAR']
    sum_scores += pitcher_dict[pitcher]['Final_Score']
    #print(pitcher, ":",pitcher_dict[pitcher])
    
avg_scores = sum_scores/len(pitch_stats)
print(avg_scores)

13.881026785714289


In [24]:
# Since we will be choosing 7 pitchers to play (5 starters and 2 relievers - Not including the bench) we should 
# naturally break our 250 dollar budget up evenly as to make an evenly balanced team

Budget = 250
Pitcher_budget = 250* (7/18) # $97.222 => $13.888 per pitcher 

final_pitchers_dict = OrderedDict(sorted(pitcher_dict.items(), key=lambda x: x[1]['Final_Score'], reverse=True))
#print(final_pitchers_dict)

In [25]:
# Print top ten picks for pitchers
top = 0
for pitcher in final_pitchers_dict:
    print(pitcher, ":")
    print("MAX Amount we will Pay: $", final_pitchers_dict[pitcher]['Final_Score'], "\n")

Jacob deGrom :
MAX Amount we will Pay: $ 94.0725 

Max Scherzer :
MAX Amount we will Pay: $ 83.26350000000001 

Justin Verlander :
MAX Amount we will Pay: $ 79.5265 

Patrick Corbin :
MAX Amount we will Pay: $ 77.921 

Gerrit Cole :
MAX Amount we will Pay: $ 75.3275 

Aaron Nola :
MAX Amount we will Pay: $ 73.4165 

Corey Kluber :
MAX Amount we will Pay: $ 72.3045 

Trevor Bauer :
MAX Amount we will Pay: $ 70.759 

Chris Sale :
MAX Amount we will Pay: $ 70.50800000000001 

Luis Severino :
MAX Amount we will Pay: $ 68.248 

Carlos Carrasco :
MAX Amount we will Pay: $ 67.299 

Miles Mikolas :
MAX Amount we will Pay: $ 62.545500000000004 

German Marquez :
MAX Amount we will Pay: $ 61.6195 

Blake Snell :
MAX Amount we will Pay: $ 61.332499999999996 

Mike Clevinger :
MAX Amount we will Pay: $ 60.7325 

Kyle Freeland :
MAX Amount we will Pay: $ 59.070499999999996 

Zack Greinke :
MAX Amount we will Pay: $ 58.89000000000001 

Dallas Keuchel :
MAX Amount we will Pay: $ 58.07150000000001 

Z


Corbin Burnes :
MAX Amount we will Pay: $ 10.105500000000001 

Jake Faria :
MAX Amount we will Pay: $ 10.0 

Taylor Cole :
MAX Amount we will Pay: $ 9.974 

Yefry Ramirez :
MAX Amount we will Pay: $ 9.962000000000002 

Tim Mayza :
MAX Amount we will Pay: $ 9.924 

Doug Fister :
MAX Amount we will Pay: $ 9.918500000000002 

Addison Reed :
MAX Amount we will Pay: $ 9.913 

Luis Perdomo :
MAX Amount we will Pay: $ 9.873999999999999 

Johnny Cueto :
MAX Amount we will Pay: $ 9.787 

Nick Tropeano :
MAX Amount we will Pay: $ 9.7315 

Trevor May :
MAX Amount we will Pay: $ 9.7175 

Ty Buttrey :
MAX Amount we will Pay: $ 9.7045 

Dan Otero :
MAX Amount we will Pay: $ 9.663 

Drew Pomeranz :
MAX Amount we will Pay: $ 9.663 

Erik Goeddel :
MAX Amount we will Pay: $ 9.655500000000002 

Sam Tuivailala :
MAX Amount we will Pay: $ 9.637 

Silvino Bracho :
MAX Amount we will Pay: $ 9.5925 

Martin Perez :
MAX Amount we will Pay: $ 9.501000000000001 

Joe Biagini :
MAX Amount we will Pay: $ 9.42600

MAX Amount we will Pay: $ -1.3555000000000001 

Willians Astudillo :
MAX Amount we will Pay: $ -1.3685 

Cody Carroll :
MAX Amount we will Pay: $ -1.3925 

Enny Romero :
MAX Amount we will Pay: $ -1.424 

Ervin Santana :
MAX Amount we will Pay: $ -1.4425 

Chris Flexen :
MAX Amount we will Pay: $ -1.4805000000000001 

Nestor Cortes :
MAX Amount we will Pay: $ -1.5055 

Eric Stout :
MAX Amount we will Pay: $ -1.512 

Robby Scott :
MAX Amount we will Pay: $ -1.524 

Jeff Beliveau :
MAX Amount we will Pay: $ -1.6555 

Gerson Bautista :
MAX Amount we will Pay: $ -1.6805 

Johnny Barbato :
MAX Amount we will Pay: $ -1.9740000000000002 

Chris Gimenez :
MAX Amount we will Pay: $ -2.037 

Jose Reyes :
MAX Amount we will Pay: $ -2.137 

Keury Mella :
MAX Amount we will Pay: $ -2.149 

Austen Williams :
MAX Amount we will Pay: $ -2.324 

JC Ramirez :
MAX Amount we will Pay: $ -2.324 

Preston Guilmet :
MAX Amount we will Pay: $ -2.474 

Blaine Boyer :
MAX Amount we will Pay: $ -2.5165 

Evan Ph

In [26]:
# Quick lookup for pitchers during the draft

print(final_pitchers_dict['Justin Verlander'])

OrderedDict([('IP', 214.0), ('K', 290.0), ('BB', 37.0), ('HBP', 8.0), ('HR', 28.0), ('SV', 0.0), ('HLD', 0.0), ('WAR', 6.9), ('Fantasy_Score', 1151.0), ('Final_Score', 79.5265)])


## Batting Selection Code ##

In [27]:
from pybaseball import batting_stats

In [28]:
batting_stats = batting_stats(2018)
batting_stats.head(10)

Season              Name       Team   Age      G     AB     PA      H  \
19  2018.0      Mookie Betts    Red Sox  25.0  136.0  520.0  614.0  180.0   
22  2018.0        Mike Trout     Angels  26.0  140.0  471.0  608.0  147.0   
43  2018.0      Jose Ramirez    Indians  25.0  157.0  578.0  698.0  156.0   
32  2018.0  Christian Yelich    Brewers  26.0  147.0  574.0  651.0  187.0   
40  2018.0      Alex Bregman     Astros  24.0  157.0  594.0  705.0  170.0   
71  2018.0  Francisco Lindor    Indians  24.0  158.0  661.0  745.0  183.0   
69  2018.0      Matt Chapman  Athletics  25.0  145.0  547.0  616.0  152.0   
53  2018.0    Anthony Rendon  Nationals  28.0  136.0  529.0  597.0  163.0   
60  2018.0     Manny Machado      - - -  25.0  162.0  632.0  709.0  188.0   
31  2018.0     J.D. Martinez    Red Sox  30.0  150.0  569.0  649.0  188.0   

       1B    2B    ...      wSL/C (pi)  wXX/C (pi)  O-Swing% (pi)  \
19   96.0  47.0    ...            1.93         NaN          0.185   
22   80.0  24.0    ...            2.59       -5.76          0.198   
43   75.0  38.0    ...            0.83         NaN          0.215   
32  110.0  34.0    ...            2.31         NaN          0.282   
40   87.0  51.0    ...            2.31        3.58          0.181   
71  101.0  42.0    ...            0.85        6.35          0.322   
69   80.0  42.0    ...            0.31         NaN          0.235   
53   93.0  44.0    ...            0.53         NaN          0.285   
60  113.0  35.0    ...            0.77         NaN          0.304   
31  106.0  37.0    ...            0.74       -3.99          0.318   

    Z-Swing% (pi)  Swing% (pi)  O-Contact% (pi)  Z-Contact% (pi)  \
19          0.535        0.356            0.658            0.926   
22          0.555        0.376            0.660            0.903   
43          0.580        0.386            0.777            0.915   
32          0.615        0.440            0.564            0.901   
40          0.568        0.371            0.781            0.920   
71          0.640        0.469            0.728            0.913   
69          0.589        0.412            0.596            0.862   
53          0.660        0.478            0.782            0.915   
60          0.713        0.487            0.637            0.885   
31          0.736        0.503            0.528            0.849   

    Contact% (pi)  Zone% (pi)  Pace (pi)  
19          0.855       0.490       23.3  
22          0.839       0.498       23.3  
43          0.874       0.468       22.9  
32          0.788       0.475       24.5  
40          0.885       0.491       25.8  
71          0.844       0.460       23.0  
69          0.787       0.501       22.2  
53          0.876       0.514       25.2  
60          0.799       0.448       25.0  
31          0.736       0.443       27.5  

[10 rows x 287 columns]

In [104]:
def returnValue(pos, value):
    return pos[value]

pd.options.display.max_rows = 4000
field = fielding()
playerData = master()
result = pd.merge(field, playerData, on='playerID', how='inner')

result = result[["playerID","nameFirst", "nameLast", "POS"]]
result['Name'] = result['nameFirst'] + ' ' + result['nameLast']

result = pd.merge(result, 
                 batting_stats,
                 on=['Name'], how='inner')

result['Fantasy_Score'] = -1*result['AB'] + 5.6*result['H'] + 2.9*result['2B'] + 5.7*result['3B'] + 9.4*result['HR'] + 3*result['BB'] + 3*result['HBP']+ 1.9*result['SB']+ -2.8*result['CS']

values = result.groupby('POS')['Fantasy_Score'].mean()

print(values)

POS
1B    387.384752
2B    346.529205
3B    375.516491
C     240.656167
OF    419.423899
P      31.484029
SS    400.066059
Name: Fantasy_Score, dtype: float64


In [83]:
batting_dict = OrderedDict()

for index, row in result.iterrows():
    batting_dict[row['Name']] = OrderedDict()
    batting_dict[row['Name']]['AB'] = row['AB']
    batting_dict[row['Name']]['H'] = row['H']
    batting_dict[row['Name']]['2B'] = row['2B']
    batting_dict[row['Name']]['3B'] = row['3B']
    batting_dict[row['Name']]['HR'] = row['HR']
    batting_dict[row['Name']]['BB'] = row['BB']
    batting_dict[row['Name']]['HBP'] = row['HBP']
    batting_dict[row['Name']]['SB'] = row['SB']
    batting_dict[row['Name']]['CS'] = row['CS']
    batting_dict[row['Name']]['POS'] = row['POS']

In [105]:
# AB=-1; H=5.6; 2B=2.9; 3B=5.7; HR=9.4; BB=3; HBP=3; SB=1.9; CS=-2.8;
for batter in batting_dict:
    batting_dict[batter]['Fantasy_Score'] = -1*batting_dict[batter]['AB'] + 5.6*batting_dict[batter]['H'] + 2.9*batting_dict[batter]['2B'] + 5.7*batting_dict[batter]['3B'] + 9.4*batting_dict[batter]['HR'] + 3*batting_dict[batter]['BB'] + 3*batting_dict[batter]['HBP']+ 1.9*batting_dict[batter]['SB']+ -2.8*batting_dict[batter]['CS']
    

In [150]:
# Calculate final score based on applying weights to both the Fantasy_Score and how their score compares to others at their position
Fantasy_weight = 0.03
Comparitive_weight = 100
sum_scores = 0

for batter in batting_dict:
    Comparitive_multiplier = (batting_dict[batter]['Fantasy_Score'] - values[batting_dict[batter]['POS']])
    batting_dict[batter]['Final_Score'] = Fantasy_weight*batting_dict[batter]['Fantasy_Score'] + (Comparitive_multiplier/Comparitive_weight)
    sum_scores += batting_dict[batter]['Final_Score']
    
avg_scores = sum_scores/len(batting_stats)
print(avg_scores)

4.875215067740579


In [151]:
Budget = 250
Batter_budget = 250* (7/18)

final_batters_dict = OrderedDict(sorted(batting_dict.items(), key=lambda x: x[1]['Final_Score'], reverse=True))

In [152]:
# Print top picks for batters
top = 0
for batter in final_batters_dict:
    print(batter, ":")
    print("MAX Amount we will Pay: $", final_batters_dict[batter]['Final_Score'], "\n")

Jose Ramirez :
MAX Amount we will Pay: $ 47.085159711075434 

Mookie Betts :
MAX Amount we will Pay: $ 46.237761006289304 

Mike Trout :
MAX Amount we will Pay: $ 45.6937610062893 

Christian Yelich :
MAX Amount we will Pay: $ 44.0337610062893 

Alex Bregman :
MAX Amount we will Pay: $ 41.39133940774488 

Manny Machado :
MAX Amount we will Pay: $ 40.96733940774489 

Francisco Lindor :
MAX Amount we will Pay: $ 40.30733940774488 

Paul Goldschmidt :
MAX Amount we will Pay: $ 40.14615248226949 

Nolan Arenado :
MAX Amount we will Pay: $ 40.04483508771929 

Freddie Freeman :
MAX Amount we will Pay: $ 39.23815248226949 

Trevor Story :
MAX Amount we will Pay: $ 39.06733940774489 

Bryce Harper :
MAX Amount we will Pay: $ 38.19376100628931 

Matt Carpenter :
MAX Amount we will Pay: $ 37.7248350877193 

Charlie Blackmon :
MAX Amount we will Pay: $ 36.1337610062893 

Giancarlo Stanton :
MAX Amount we will Pay: $ 36.0617610062893 

Javier Baez :
MAX Amount we will Pay: $ 35.32733940774488 

Mi

MAX Amount we will Pay: $ -0.35484028892455555 

Adam Liberatore :
MAX Amount we will Pay: $ -0.35484028892455555 

Hansel Robles :
MAX Amount we will Pay: $ -0.35484028892455555 

Ryan Weber :
MAX Amount we will Pay: $ -0.35484028892455555 

Mike Wright :
MAX Amount we will Pay: $ -0.35484028892455555 

Buddy Baumann :
MAX Amount we will Pay: $ -0.35484028892455555 

Austin Brice :
MAX Amount we will Pay: $ -0.35484028892455555 

Luis Cessa :
MAX Amount we will Pay: $ -0.35484028892455555 

John Gant :
MAX Amount we will Pay: $ -0.35484028892455555 

Reynaldo Lopez :
MAX Amount we will Pay: $ -0.35484028892455555 

Adalberto Mejia :
MAX Amount we will Pay: $ -0.35484028892455555 

Wandy Peralta :
MAX Amount we will Pay: $ -0.35484028892455555 

Alex Reyes :
MAX Amount we will Pay: $ -0.35484028892455555 

Warwick Saupold :
MAX Amount we will Pay: $ -0.35484028892455555 

Julio Urias :
MAX Amount we will Pay: $ -0.35484028892455555 

Luke Weaver :
MAX Amount we will Pay: $ -0.370840288

In [128]:
# Quick lookup for batters during the draft

print(final_batters_dict['Nolan Arenado'])

OrderedDict([('AB', 590.0), ('H', 175.0), ('2B', 38.0), ('3B', 2.0), ('HR', 38.0), ('BB', 73.0), ('HBP', 3.0), ('SB', 2.0), ('CS', 2.0), ('POS', '3B'), ('Fantasy_Score', 1094.9999999999998), ('Final_Score', 69.1396701754386)])


## Writeup ##

### Introduction ###
Fantasy sports have become increasingly popular over the last decade with more and more participants becoming involved each year. With more players and larger leagues, it is difficult for individual participants to gain a competitive advantage over their adversaries using personally acquired knowledge alone. That is, routinely watching sports alone does not give fantasy sports competitors enough of an advantage to build reliable fantasy teams that consistantly out perform their opponents.  

For this assignment, we developed a set of algorithms that we believe will help us draft fantasy baseball players that will maximize our teams performance within our league. In other words, our algorithms will help us draft the ideal players for out playing our league opponents. We split these algorithms into two unique sets - one for pitchers and one for batters. 

### Data and Statistics ###
The pitching algorithms are relatively simple. We first took data from the 2018 MLB season (since this was the most recent season to present day). We then determined the fantasy score for all pitchers in this dataset. Note that a pitcher's fantasy score depended on the following statistics: 
* Innings Pitched (IP) = 5 points each
* Strikeout (K) = 2 points each
* Walk (BB) = -3 points each
* Hit By Pitch (HBP) = -3 points each
* Homerun (HR) = -13 points each
* Save (SV) = 5 points each
* Hold (HLD) = 4 points each

Once we determined each pitchers fantasy score, we then determined each pitchers Wins Above Replacement (WAR) metric using the Fangraphs evaluation for WAR. We believed that WAR was the most significant statistic for choosing a fantasy pitcher because it compares a given pitchers stats to that of an average pitcher. By fault, the pitchers with higher WARs can be viewed as better pitchers compared to their league average replacements. We then determined the pitchers' final scores by taking a weighted sum of their WAR values and their fantasy scores. Note that we set up the weight proportions so that the average final score equaled the average amount of money we had to spend for each player - \$13.88. With the weights set up this way, our pitcher selection algorithm not only determines a score for a pitcher, but it also determines a maximum amount of money to spend on said player.

Similarly, for the batting algorithms we took the data from the 2018 MLB season and used that to create the fantasy score for all of the offensive players in the dataset. For the batting algorithm we did not use the traditional WAR statisitc as defense does not matter in terms of scoring fantasy points. As it is pointed out in the assignment however, we still are going to want to be able to determine a player's skill relative to their position so that we place higher value on players who score high points from positions that are not as "deep".  A batter's fantasy score depends on the following statistics: 
* At Bats (AB) = -1
* Hits (H) = 5.6
* Doubles (2B) = 2.9
* Triples (3B) = 5.7
* Home Runs (HR) = 9.4
* Walks (BB) = 3
* Hit by Pitch (HBP) = 3
* Stolen Bases (SB) = 1.9
* Caught Stealing (CS) = -2.8

So we took the overall results from the 2018 season and then modify that number based on their relative skill level within the position and weighting to determine the maximum amount of money to spend on the player.

### Results ###
The results for the pitchers algorithm are as follows:

TOP FIVE PITCHERS:
* Jacob deGrom - MAX we will pay: \$95.02
* Max Scherzer - MAX we will pay: \$83.58
* Justin Verlander - MAX we will pay: \$79.20
* Patrick Corben - MAX we will pay: \$76.96
* Gerrit Cole - MAX we will pay: \$74.69

Note: as mentioned above the max amount that we will pay a given pitcher is equivalent to our algorithm's rating of that pitcher.

We can derive this rating by looking at Justin Verlander as an example. Verlander had the following stats in 2018:
* IP: 214
* K: 290
* BB: 37
* HBP: 8
* HR: 28
* SV: 0
* HLD: 0
* WAR: 6.8
By applying point values to each of Verlander's stats above, we were able to determine that Verlander's fantasy score equaled 1151. The tuned weights for fantasy score and WAR equaled .05 and 3.185 respectively. Therefore, Verlander's final score using our system euqaled .05*1151 + 3.185*6.8 = 79.208, which also equates to the absolute maximum amount of money we would be willing to spend on him.

TOP FIVE BATTERS:
* Jose Ramirez :
MAX Amount we will Pay: $ 47.08

* Mookie Betts :
MAX Amount we will Pay: $ 46.23

* Mike Trout :
MAX Amount we will Pay: $ 45.69

* Christian Yelich :
MAX Amount we will Pay: $ 44.03

* Alex Bregman :
MAX Amount we will Pay: $ 41.39 

### Limitations ###
One limitation of our pitching metric is that severe "skewedness" of the final metric we use for pitcher evaluation. Since some pitchers acquired massive stats in 2018, their final scores were considerably higher compared to the rest of the average pitchers. Due to this skew in the data, our top pitchers have incredibly large salary caps that could get us into trouble during the draft. That is if we go big on one of these pitchers, we will most likely have to compromise with average pitchers for the remainder of our team. Come crunch time, it may be difficult to determine which average pitchers we want to draft if we have already drafted a big name pitcher.<br>
A limitation on the batting metrics is that stats are going to skew heavily in favor or players who were not injured last season because the majority of the weight comes from their total batting numbers acrossed all of the games. So if a player is out for part of the season due to injury, or was rested or benched for some period of the season that is going to strongly impact the fantasy score that they received with our system.
